File: Next_Permutation.py
class Solution:

    # @param num, a list of integer
    # @return a list of integer
    def nextPermutation(self, num):
        if len(num) <= 1:
            return num

        for idx in range(len(num) - 2, -1, -1):
            ii = idx + 1
            if num[idx] < num[ii]:
                j = len(num) - 1
                while not num[idx] < num[j]:
                    j -= 1
                num[idx], num[j] = num[j], num[idx]
                end = len(num) - 1
                while ii < end:
                    num[ii], num[end] = num[end], num[ii]
                    ii += 1
                    end -= 1
                return num
            if idx == 0:
                num.reverse()
                return num

if __name__ == '__main__':
    s = Solution()
    lst = [1]
    while True:
        print lst
        s.nextPermutation(lst)

File: Next_Permutation.cpp
#include <vector>
using std::vector;
using std::swap;
#include <algorithm>
using std::reverse;

class Solution {
public:
    void nextPermutation(vector<int> &num) {
        __next_perm(num.begin(), num.end());
    }

private:
    bool __next_perm(vector<int>::iterator beg, vector<int>::iterator end) {
        if (beg == end) {
            return false;
        }

        vector<int>::iterator i = beg;
        ++i;
        if (i == end) {
            return false;
        }

        i = end;
        --i;

        for (;;) {
            vector<int>::iterator ii = i;
            --i;
            if (*i < *ii) {
                vector<int>::iterator j = end;
                while (!(*i < *--j));

                std::iter_swap(i, j);
                reverse(ii, end);
                return true;
            }

            if (i == beg) {
                reverse(beg, end);
                return false;
            }
        }
    }
};

int main() {
    Solution s;
    vector<int> num {1, 1, 5};
    while (true) {
        s.nextPermutation(num);
        for (auto &item : num) {
            printf("%d ", item);
        }
        printf("\n");
    }
    return 0;
}
